### ISSUE TITLE
[Feature Request] Allow parse mca file from std::io::Read

### ISSUE THREAD

**Issue by ToKiNoBug** [funnyboy-roks/mca-parser#1]
Sometimes we need to parse a mca file that is not stored in filesystem directly(for example, in a zip file), thus we have to copy the file into disk, which is not convenient. I hope to add a function like 
    `parse_mca_reader(src: &mut dyn Read, coord: RegionCoord)`
so that we can parse mca file from any source.

**Comment by funnyboy-roks** [comment]
I think that's a great idea, in fact, I'm surprised I didn't do that from the beginning.  I see you forked the repo, are you intending to create a PR for this?

**Comment by funnyboy-roks** [comment]
I do, however, think that a signature of
```rs
parse_mca_reader<R>(src: &mut R, coord: RegionCoord) where R: Read
```
is probably better so that we don't have the runtime overhead of using `dyn`.

**Comment by ToKiNoBug** [comment]
> I think that's a great idea, in fact, I'm surprised I didn't do that from the beginning. I see you forked the repo, are you intending to create a PR for this?

I tried, but I didn't come up with an idea on how to solve it. Should we save `Box<dyn Read>` as a member of `RegionParser`? By doing that, I found it hard to pass borrow checking.

**Comment by ToKiNoBug** [comment]
> I do, however, think that a signature of
> 
> ```rust
> parse_mca_reader<R>(src: &mut R, coord: RegionCoord) where R: Read
> ```
> 
> is probably better so that we don't have the runtime overhead of using `dyn`.

That would be perfect.

**Comment by funnyboy-roks** [comment]
I took a look into the code a bit and it would require a decent amount of reorganisation in order to work in a nice way.  From what I can tell, there are two ways we could make this work:
- `RegionParser` owns the reader or
- the reader is passed in for every method call

The former would be a little easier from the side of implementing it, I think, but the latter is better for the user since they don't have to give away ownership and they can manage where the reader is borrowed.

Because of this, the ideal thing (in my mind) would be to remove the `path` field from the `RegionParser` struct and instead pass the reader through the functions, but this would present difficulty with functions like `from_directory`, which rely on the `RegionParser` being able to know from which file it is reading.

A potential solution to this problem would be to make the `Dimension` struct (returned from `from_directory`) use a struct of some kind that contains the file path and `RegionParser` and has a method similar to the `RegionParser`'s `parse` that would open the file and call `RegionParse::parse`.

In summary, the changes that would need to occur are: 
- Remove the `path` field from `RegionParser`
- make `parse_data_with_option` accept a generic `R` where `R: Read`
- make `parse` accept a generic `R` where `R: Read`
- make `parse_without_data` accept a generic `R` where `R: Read`
- make `parse_chunks` accept a generic `R` where `R: Read`
- Update `from_file` to open the file
- Create a struct which can hold a `PathBuf` and `RegionParser` (maybe something like `FileRegionParser`)
- Update `Dimension` to use this struct instead of `RegionParser` in its `regions` HashMap.

**Comment by funnyboy-roks** [comment]
I was looking at the library and decided that I didn't really like it at all, so I decided to rewrite it.  The rewrite now does everything it did before, but it is much nicer and has support for creating a region from readers, paths, and a custom trait that the user can implement.

I'll publish it later today in place of the current library.
